api/dbadapters/bigquery.ts

```diff
@@ -257,54 +248,82 @@ export class BigQueryDbAdapter implements IDbAdapter {
-            throw new Error(`Could not find table metadata for ${JSON.stringify(transitiveInput)}`);
-          }
-          return tableMetadataByTarget.get(transitiveInput);
-        })
-      });
-
-      const targetName = `${action.target.database}.${action.target.schema}.${action.target.name}`;
-
-      return buildQuery(targetName, persistTable);
-    });
+    const targetStringifier = JSONObjectStringifier.create<dataform.ITarget>();
+    try {
+      // Create the cache table, if needed.
+      await this.execute(
+        `
+CREATE TABLE IF NOT EXISTS \`${CACHED_STATE_TABLE_NAME}\` (
+  target STRING,
+  metadata_proto STRING
+)`,
+        options
+      );
+      // Before saving any new data, delete all entries for 'allActions'.
+      await this.execute(
+        `
+DELETE \`${CACHED_STATE_TABLE_NAME}\` WHERE target IN (${allActions
+          .map(({ target }) => `'${targetStringifier.stringify(target)}'`)
+          .join(",")})`,
+        options
+      );
 
-    const unionQuery = queries.join(" UNION ALL ");
-
-    const updateQuery = `MERGE INTO \`${CACHED_STATE_TABLE_NAME}\` T
-    USING (${unionQuery}) S
-    ON (T.target_name = S.target_name)
-    WHEN NOT MATCHED THEN
-      INSERT (target_name, metadata_json, metadata_proto)
-      VALUES(S.target_name, S.metadata_json, S.metadata_proto)
-    WHEN MATCHED THEN
-      UPDATE SET metadata_json = S.metadata_json,
-          metadata_proto = S.metadata_proto;`;
-    await this.runQuery(updateQuery);
+      // Save entries for 'actionsToPersist'.
+      const transitiveInputMetadataByTarget = await this.computeTransitiveInputMetadataMap(
+        actionsToPersist
+      );
+      const valuesTuples = actionsToPersist
+        // If we were unable to load metadata for the action's output dataset, or for any of the action's
+        // input datasets, do not store a cache entry for the action.
+        .filter(
+          action =>
+            transitiveInputMetadataByTarget.has(action.target) &&
+            action.transitiveInputs.every(transitiveInput =>
+              transitiveInputMetadataByTarget.has(transitiveInput)
+            )
+        )
+        .map(
+          action =>
+            `('${targetStringifier.stringify(action.target)}', '${encodePersistedTableMetadata({
+              target: action.target,
+              lastUpdatedMillis: transitiveInputMetadataByTarget.get(action.target)
+                .lastUpdatedMillis,
+              definitionHash: hashExecutionAction(action),
+              transitiveInputTables: action.transitiveInputs.map(transitiveInput =>
+                transitiveInputMetadataByTarget.get(transitiveInput)
+              )
+            })}')`
+        );
+      // We have to split up the INSERT queries to get around BigQuery's query length limit.
+      while (valuesTuples.length > 0) {
+        let valuesToInsert = valuesTuples.pop();
+        while (valuesTuples.length > 0 && valuesToInsert.length < MAX_QUERY_LENGTH * 0.95) {
```

Comment:
If this ever goes wrong, I imagine it will cause us a major headache. Seems like it wouldn't be hard to change this code to compute size iff we added the next set of values to the query and unpop (or just peek instead of popping immediately).